### ISSUE TITLE
Apollo 3.0.0-beta2 subscriptions do not collects

### ISSUE THREAD

**Issue by simondorociak** [apollographql/apollo-android#3583]
I have this in Fragment and subscription do not collect:

```
lifecycleScope.launch {
            client.subscribe(<my_subscription>).execute()
                .flowWithLifecycle(lifecycle, Lifecycle.State.STARTED)
                .retryWhen { cause, _ ->
                    cause.printStackTrace()
                    delay(10000L)
                    true
                }
                .collect {
                    Timber.d("onNewMessage invoked")
                }
        }
```

 Timber.d("onNewMessage invoked") is never invoked. Can you tell why this happen? Thanks.

- Im using 3.0.0-beta2 version of library.

**Comment by martinbonnin** [comment]
Hi 👋 Can you share your `ApolloClient` construction code? It might be that subscriptions require a different url? 

```
val apolloClient = ApolloClient.Builder()
  .serverUrl("https://example.com/graphql")
  .webSocketServerUrl("wss://example.com/subscriptions")
  .build()
```

See [here](https://www.apollographql.com/docs/android/v3/essentials/subscriptions/) for more details.

**Comment by simondorociak** [comment]
I will send you asap thanks!

S pozdravom / Best regards

Ing. Šimon Dorociak
Android Engineer
Tel: +420 776 226 886
________________________________

**Comment by simondorociak** [comment]
> Hi 👋 Can you share your `ApolloClient` construction code? It might be that subscriptions require a different url?
> 
> ```
> val apolloClient = ApolloClient.Builder()
>   .serverUrl("https://example.com/graphql")
>   .webSocketServerUrl("wss://example.com/subscriptions")
>   .build()
> ```
> 
> See [here](https://www.apollographql.com/docs/android/v3/essentials/subscriptions/) for more details.

Im configuring ApolloClient like this:

```
@Provides
    @Singleton
    fun provideApolloClient(
            application: Application,
            prefs: SharedPrefsManager,
            httpClient: OkHttpClient) : ApolloClient {
        return ApolloClient.Builder()
            .networkTransport(
                HttpNetworkTransport(
                    serverUrl = BuildConfig.SERVER_URL,
                    okHttpClient = httpClient
                )
            )
            .subscriptionNetworkTransport(
                AuthSocketSubscriptionTransport(
                    prefs,
                    BuildConfig.SOCKET_URL,
                    DefaultWebSocketEngine(httpClient)
                )
            )
            .addInterceptor(RefreshTokenInterceptor(application, prefs))
            .build()
    }

I dont see function webSocketServerUrl in my project.

```
 In my case I have my own AuthSocketSubscriptionTransport and also WsProtocol because I have to pass JWT token in payload but this I just copy & paste your default subs. transport and I just modified piece of code where Im puting JWT token into payload:

```
class AuthSubscriptionWsProtocol(
    prefs: SharedPrefsManager,
    webSocketConnection: WebSocketConnection,
    listener: Listener,
    private val connectionAcknowledgeTimeoutMs: Long = 10_000,
    private val connectionPayload: suspend () -> Map<String, Any?>? = { mapOf("Authorization" to  "JWT ${prefs.getJwtToken()}") },
) : WsProtocol(webSocketConnection, listener)
```

Our socket url is classic : wss:///<domain>/graphql/. 

It worked on version 2 but version 2 had some bugs which were for us important so we switched to version 3 because we want to use new technologies as possible.

I checked code, added some logging and I see in logcat that is called this function in our AuthSocketSubscriptionTransport:

```
override fun <D : Operation.Data> execute(
        request: ApolloRequest<D>,
    ): { ... }
```

This normally pass through filter and transformWhile functions and also with map but in Fragment i dont receive event from socket in collect callback.

**Comment by simondorociak** [comment]
@martinbonnin  I fixed issue! It was caused by my incorrect implementation of ApolloInterceptor.  I will write my usercase maybe it will help someone else to fix issue.

Scenario: Apollo doesn't have component called "Authenticator" which is used by okhttp it works for REST but not for GraphQL. So. in order to achieve this behaviour I implemented custom ApolloInterceptor in which Im executing request by:

`chain.proceed(request).single()`

and then checking for errors and exact 401 HTTP code. This works well but that single() function someone changes also request of subscription and this caused that I did not receive collect callbacks. 

So I added condition that if I'm intercepting subscription call i will just pass it with proceed and intercept all other http requests (operations).

```
val operationName = request.operation.name()
if (operationName.contains(OPERATION_SUBSCRIPTION, true)) {
   Timber.d("Subscription request detected -> do not intercept")
      return chain.proceed(request)
} else {
   // intercept other requests
}
```

Thank you for your time.
